### PR TITLE
Partially fix #3160 by properly handling resultsFormat query

### LINKED ISSUE #3160: 🐛 BUG: when trying to left join 2 tables that have same name columns - second column is not returned
### Which Cloudflare product(s) does this pertain to?

D1

### What version of `Wrangler` are you using?

2.19.0

### What operating system are you using?

macOS

### Describe the Bug

As per issue [#555](https://github.com/drizzle-team/drizzle-orm/issues/555) we can see, that d1 stopped returning duplicate names from different tables. It was working before and it's how any other driver is doing for `raw` mode

```bash
$ pnpm wrangler d1 execute dzltest --command='select "users"."id", "posts"."id", "posts"."ownerId", "posts"."title", "posts"."content" from "users" left join "posts" on "posts"."ownerId" = "users"."id";'

┌────┬─────────┬───────┬───────────┐
│ id │ ownerId │ title │ content   │
├────┼─────────┼───────┼───────────┤
│ 2  │ 1       │ test  │ 111111111 │
├────┼─────────┼───────┼───────────┤
│ 3  │ 1       │ ueoau │ ueueu     │
├────┼─────────┼───────┼───────────┤
│ 1  │ 1       │ xxx   │ yyy       │
└────┴─────────┴───────┴───────────┘
```

But should return `id` for user and `id` for posts as well
Expected
```bash
┌────┬────┬─────────┬───────┬───────────┐
│ id │ id │ ownerId │ title │ content   │
├────┼────┼─────────┼───────┼───────────┤
│ 2  │ 1  │    1    │ test  │ 111111111 │
├────┼────┼─────────┼───────┼───────────┤
│ 3  │ 2  │    1    │ ueoau │ ueueu     │
├────┼────┼─────────┼───────┼───────────┤
│ 1  │ 3  │    1    │ xxx   │ yyy       │
└────┴────┴─────────┴───────┴───────────┘
```

### PR DESCRIPTION
**Pull request by kossnocorp**
TL;DR: Cloudflare D1 has a critical bug in Workers SDK reproducible on the Cloudflare D1 Console and when running D1 through Wrangler. The PR fixes the bug and also improves compatibility with `workerd`.

### The problem

The essence of the problem is selecting two columns with similar names:

```sql
SELECT
    projects.name,
    organizations.name
FROM
    projects
    JOIN organizations ON projects.organization_id = organizations.id;
```

...must return in the row containing both columns (tested on PostgreSQL v16.3):

```
test=# SELECT projects.name, organizations.name
        FROM projects
        JOIN organizations ON projects.organization_id = organizations.id;
    name    |    name    
------------+------------
 Cloudflare | Cloudflare
(1 row)
```

SQLite (on a database created by Wrangler) gives a similar result:

```
sqlite> SELECT
    projects.name,
    organizations.name
FROM
    projects
    JOIN organizations ON projects.organization_id = organizations.id;
Daisy Chain|Daisy Chain
Fake Daisy Chain|Fake Org
Mind Control|Daisy Chain
sqlite> 
```

I also [added a test](https://github.com/cloudflare/workerd/pull/2166) for this behavior to [workerd](https://github.com/cloudflare/workerd) that is passing:

```ts
await itShould(
  // ...
  () =>
    DB.prepare(
      `SELECT projects.name, organizations.name
      FROM projects
      JOIN organizations ON projects.organization_id = organizations.id;`
    ).raw(),
  [['Cloudflare', 'Cloudflare']]
)
```

However, when sending the same request through Workers SDK, I get:

```
| name       |
| Cloudflare |
```

As both columns have the same name, `name`, they got squashed into a single column. 

A GitHub user [raised the issue a year ago](https://github.com/cloudflare/workers-sdk/issues/3160) and since then it [got fixed in `workerd`](https://github.com/cloudflare/workers-sdk/issues/3160#issuecomment-1915805490) ([that my test confirms](https://github.com/cloudflare/workerd/pull/2166)), but it is still present in Workers SDK.

The problem is also reproducible in Cloudflare D1 Console, where I get the same data:

```
| name       |
| Cloudflare |
```

It [breaks people's code in unexpected ways](https://github.com/drizzle-team/drizzle-orm/issues/555) and creates an impression of D1 as an unreliable product that might behave differently in production.

In my opinion, it's a critical bug that must be prioritized.

### The source

I chased the bug through `wrangler` → `workerd` → `miniflare` [to these line](https://github.com/cloudflare/workers-sdk/blob/81dfb1746a2b2a17c06f809b2da9f937810ca701/packages/miniflare/src/workers/d1/database.worker.ts#L123-L124) in `database.worker.ts`:

```ts
const cursor = this.db.prepare(query.sql)(...params);
const results = convertResults(all(cursor));
```

When running [the test](https://github.com/cloudflare/workers-sdk/pull/5917/commits/022971e9299d25e1e4eb183434a6b1a3779b3f2c#diff-e0656c74bcb20ddb6d1bcec2603532d76a82cbe87475aa3efecf13c5c05b1064R465-R481):

```ts
await db
  .prepare(
	`SELECT ${tableColours}.name, ${tablePalettes}.name FROM ${tableColours} JOIN ${tablePalettes} ON ${tableColours}.id = ${tablePalettes}.colour_id`
  )
  .raw();
```

...the `cursor` object passed through `all` results in the given structure:

```
[{ name: "Night" }]
```

You can see that at this point, the information is already lost, and the `name` columns have been reduced to one property.

After further researching I found that [`workerd` sends the `resultsFormat` query](https://github.com/cloudflare/workerd/blob/1d89f3b8e9cdcd898ea486656d72d9551e79f4a3/src/cloudflare/internal/d1-api.ts#L319) that determines the format of the `results`.

It worked before only because `workerd` handles the case when the API always returns an array of objects either by sheer luck or as a workaround for Workers SDK behavior.

### The solution

The PR fixes the problem and adds proper handling of the `resultsFormat` query.

It makes `raw` preserve columns with the same name, addressing the original problem.

It also adds handling of `NONE` which `workerd` sends with [`run`](https://github.com/cloudflare/workerd/blob/1d89f3b8e9cdcd898ea486656d72d9551e79f4a3/src/cloudflare/internal/d1-api.ts#L295) and [`exec`](https://github.com/cloudflare/workerd/blob/1d89f3b8e9cdcd898ea486656d72d9551e79f4a3/src/cloudflare/internal/d1-api.ts#L103) (hence the change in existing test cases).

### Extras

Here's the SQL I used to test the behavior:

```sql
CREATE TABLE organizations (id INTEGER PRIMARY KEY, name TEXT);

CREATE TABLE projects (
    id INTEGER PRIMARY KEY,
    name TEXT,
    organization_id INTEGER,
    FOREIGN KEY (organization_id) REFERENCES organizations(id)
);

INSERT INTO
    organizations (id, name)
VALUES
    (1, 'Cloudflare');
    
INSERT INTO
    projects (id, name, organization_id)
VALUES
    (1, 'Cloudflare', 1);
    
SELECT
    projects.name,
    organizations.name
FROM
    projects
    JOIN organizations ON projects.organization_id = organizations.id;
```

## What this PR solves / how to test

The PR fixes #3160.

## Author has addressed the following

- Tests
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- E2E Tests CI Job required? (Use "e2e" label or ask maintainer to run separately)
  - [x] I don't know
  - [ ] Required / Maybe required
  - [ ] Not required because:
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- Public documentation
  - [ ] TODO (before merge)
  - [ ] Cloudflare docs PR(s): <https://github.com/cloudflare/cloudflare-docs/pull/>...
  - [x] Not necessary because: it is expected behavior

### CODE REVIEW

**Comment by kossnocorp** [comment]
@geelen @petebacondarwin please take a look again. I found the root cause of the problem and fixed the issue.

I bealive it's a critical bug as it [breaks people code in unexpected way](https://github.com/drizzle-team/drizzle-orm/issues/555) and creates an impression of D1 as unreliable product that might behave differently in production.

I'm confident in the fix quality, so I think it won't create much work for you and fix an important problem for customers.

File: packages/miniflare/src/workers/d1/database.worker.ts [hunk 1]
```diff
@@ -113,15 +132,20 @@ export class D1DatabaseObject extends MiniflareDurableObject {
...
-	#query = (query: D1Query): D1SuccessResponse => {
+	#query = (format: D1ResultsFormat, query: D1Query): D1SuccessResponse => {
 		const beforeTime = performance.now();
 
 		const beforeSize = this.state.storage.sql.databaseSize;
 		const beforeChanges = this.#changes();
 
 		const params = convertParams(query.params ?? []);
 		const cursor = this.db.prepare(query.sql)(...params);
-		const results = convertResults(all(cursor));
```
Comment (rozenmd) [inline packages/miniflare/src/workers/d1/database.worker.ts:124]: Looks like you'd need to remove the import for `all`

Comment (kossnocorp) [reply packages/miniflare/src/workers/d1/database.worker.ts:124]: Done!

**Comment by rozenmd** [comment]
Sorry, looks like the build is failing - can you run prettier against `packages/miniflare/test/plugins/d1/test.ts`?

File: packages/miniflare/test/plugins/d1/test.ts [hunk 2]
```diff
@@ -18,6 +18,7 @@ export interface Context extends MiniflareTestContext {
 	db: D1Database;
 	tableColours: string;
 	tableKitchenSink: string;
+	tablePalettes: string
```
Comment (rozenmd) [inline packages/miniflare/test/plugins/d1/test.ts:21]:
```suggestion
	tablePalettes: string;
```

**Comment by rozenmd** [comment]
nvm found the issue, missing `;`

**Comment by rozenmd** [comment]
That's alright @kossnocorp - I can fix it up

**Comment by kossnocorp** [comment]
@rozenmd thank you, I appreciate it!

**Comment by rozenmd** [comment]
Looked at this a bit more - can you remove this part?

> It also adds handling of NONE which workerd sends with [run](https://github.com/cloudflare/workerd/blob/1d89f3b8e9cdcd898ea486656d72d9551e79f4a3/src/cloudflare/internal/d1-api.ts#L295) and [exec](https://github.com/cloudflare/workerd/blob/1d89f3b8e9cdcd898ea486656d72d9551e79f4a3/src/cloudflare/internal/d1-api.ts#L103) (hence the change in existing test cases).

We initially released this change in workerd and had to rapidly roll-forward a fix in our D1 Worker to prevent breaking compatibility with user Workers that relied on `run`/`exec` to return results. We intend to fix this properly via a `compatibility_date` later.

**Comment by kossnocorp** [comment]
@rozenmd sure, I'm on it.

**Comment by kossnocorp** [comment]
@rozenmd done!

File: packages/miniflare/test/plugins/d1/suite.ts [hunk 3]
```diff
@@ -392,10 +403,29 @@ test("D1PreparedStatement: raw", async (t) => {
 		.bind("yellow")
 		.first("id");
 	t.is(id, 4);
+
+	// Check whether workerd raw test case passes here too
+	// Note that this test did not pass with the old binding
+	if (!t.context.bindings["__D1_BETA__DB"]) {
```
Comment (rozenmd) [inline packages/miniflare/test/plugins/d1/suite.ts:409]: This test suite is dynamic (yay) and gets run by both versions of the D1 binding.

I had to add this check to ensure that it's fixed only for new bindings (since old versions of wrangler have the old binding)

File: .changeset/little-jars-train.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+---
+"miniflare": minor
+---
+
+Fix JOIN behaviour when selecting columns with the same name.
```
Comment (CarmenPopoviciu) [inline .changeset/little-jars-train.md:5]: can we pls use the recommended [changesets format](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changeset-message-format) 🙏

Comment (CarmenPopoviciu) [inline .changeset/little-jars-train.md:5]:
```suggestion
fix: Fix JOIN behaviour when selecting columns with the same name.
```

Comment (rozenmd) [inline .changeset/little-jars-train.md:5]:
```suggestion
fix: D1's JOIN behaviour when selecting columns with the same name.
```

File: .changeset/little-jars-train.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+---
+"miniflare": minor
+---
+
+fix: D1's JOIN behaviour when selecting columns with the same name.
+
+Properly handle the `resultsFormat` query that `workerd` sends. This fixes [the JOIN bug](https://github.com/cloudflare/workers-sdk/issues/3160) and makes the behaviour of `raw` consistent with the `workerd` behaviour.
```
Comment (rozenmd) [inline .changeset/little-jars-train.md:7]:
```suggestion
Properly handle the `resultsFormat` query that `workerd` sends. This fixes [the JOIN bug](https://github.com/cloudflare/workers-sdk/issues/3160) and makes the behaviour of `raw` consistent with the `workerd` behaviour.

Closes https://github.com/cloudflare/workers-sdk/issues/3160
```

**Review by CarmenPopoviciu — APPROVED** [review]

File: .changeset/little-jars-train.md [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+
+Closes https://github.com/cloudflare/workers-sdk/issues/3160
```
Comment (rozenmd) [inline .changeset/little-jars-train.md:9]:
```suggestion
```

File: .changeset/little-jars-train.md [hunk 7]
```diff
@@ -0,0 +1,8 @@
+---
+"miniflare": minor
+---
+
+fix: D1's JOIN behaviour when selecting columns with the same name.
+
+Properly handle the `resultsFormat` query that `workerd` sends. This fixes [the JOIN bug](https://github.com/cloudflare/workers-sdk/issues/3160) and makes the behaviour of `raw` consistent with the `workerd` behaviour.
```
Comment (rozenmd) [inline .changeset/little-jars-train.md:7]:
```suggestion
Properly handle the `resultsFormat` query that `workerd` sends. This partially fixes [the JOIN bug](https://github.com/cloudflare/workers-sdk/issues/3160) and makes the behaviour of `raw` consistent with the `workerd` behaviour.
```

File: packages/miniflare/test/plugins/d1/suite.ts (hunk 3 above)
Comment (kossnocorp) [reply packages/miniflare/test/plugins/d1/suite.ts:409]: Fascinating!

**Comment by kossnocorp** [comment]
@rozenmd thank you for helping me to land this PR into the main branch 🙏
